contracts/cw721-base/src/contract.rs

```diff
@@ -14,442 +12,609 @@ use cw721::{
 
 use crate::error::ContractError;
 use crate::msg::{ExecuteMsg, InstantiateMsg, MintMsg, MinterResponse, QueryMsg};
-use crate::state::{
-    increment_tokens, num_tokens, tokens, Approval, TokenInfo, CONTRACT_INFO, MINTER, OPERATORS,
-};
-use cw_storage_plus::Bound;
+use crate::state::{token_owner_idx, Approval, TokenIndexes, TokenInfo};
+use cw_storage_plus::{Bound, IndexedMap, Item, Map, MultiIndex};
+use schemars::JsonSchema;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::marker::PhantomData;
 
 // version info for migration info
 const CONTRACT_NAME: &str = "crates.io:cw721-base";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
-#[cfg_attr(not(feature = "library"), entry_point)]
-pub fn instantiate(
-    deps: DepsMut,
-    _env: Env,
-    _info: MessageInfo,
-    msg: InstantiateMsg,
-) -> StdResult<Response> {
-    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
-
-    let info = ContractInfoResponse {
-        name: msg.name,
-        symbol: msg.symbol,
-    };
-    CONTRACT_INFO.save(deps.storage, &info)?;
-    let minter = deps.api.addr_validate(&msg.minter)?;
-    MINTER.save(deps.storage, &minter)?;
-    Ok(Response::default())
+const DEFAULT_LIMIT: u32 = 10;
+const MAX_LIMIT: u32 = 30;
+
+pub struct Cw721Contract<'a, T, C>
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    pub contract_info: Item<'a, ContractInfoResponse>,
+    pub minter: Item<'a, Addr>,
+    pub token_count: Item<'a, u64>,
+    /// Stored as (granter, operator) giving operator full control over granter's account
+    pub operators: Map<'a, (&'a Addr, &'a Addr), Expiration>,
+    pub tokens: IndexedMap<'a, &'a str, TokenInfo<T>, TokenIndexes<'a, T>>,
+
+    _custom_response: PhantomData<C>,
```

Comment:
There is no reason to make this contract generic over `C`. If contract interface would be a trait, it should be an associated type (as for any type `T` there is only one well defined type `C`).

Reply:
Good point. Each method (`execute_*`) should then just take `C` as a generic parameter for this method, right?

I didn't think of it, but also part of it comes from laziness.

`C: Clone + std::fmt::Debug + PartialEq + JsonSchema` is a mouthful to put everywhere.

I would like to do something like `type CustomMsg = Clone + std::fmt::Debug + PartialEq + JsonSchema` to make this simpler, but last time I tried something similar, I got compiler errors. I could make a trait:

```rust
pub trait CustomMsg: Clone + std::fmt::Debug + PartialEq + JsonSchema {}
```

But that would require me to also implement the trait specifically for all messages (which are very few, so many that is actually a good idea)

Reply:
@hashedone I addressed this in b32b21e6

I agree it is more proper. I also made the CustomMsg trait as a placeholder (but we need to move it somewhere else to really use it).

One thing I noticed is it made tests a bit more verbose (adding `::<Empty>` in lots of places), but this is the cost of generics I guess.

Reply:
The issue is that I want my `Cw721BaseContract` struct to implement` Cw721<C, T>` for any `C` and `T`.
There is no correlation between the two. C is usually Empty, but if you need to integrate with Custom blockchain messages it is possible. `T` may be many values based on who is extending it.

If I do something like:

```rust
trait Cw721<T> where T: Serialize + DeserializeOwned + Clone {
  type Err: ToString;
  type C: Serialize + DeserializeOwned + Clone;
}
```

like you do below, it hardcodes the Response<C> and only allows 1 C for a given contract. This does now allow us to extend it for CustomMsg.

I agree fully with `type Err`, as that is fixed in the implementation (but not by the trait). But I would move C up to a trait generic rather than a trait attribute (not sure the proper terms here... please correct me, but I think you got the idea)

Reply:
I am looking at this and I don't get the usecase, but it would probably be more clear when I would see it. I just cannot imagine a case when you want more than one implementation of a trait per contract. Now I am even thinking that even T is probably the associated type, as at the end for your entry points you have to make proper type for deserialization, and at the end your contract is basically setting it in stone. On the other hand it may be, that the contract itself is operating on very specific `T`/`C`, but maybe somewhere in future when you extend it you may pass-forward calls and then it is useable. But I don't see generic usage in C in the implementation so I am not sure.

Anyway - let leave it for now, it is huge step forward anyway, and I think if we will observe how it evolves, we would have better POV on this.